Implement RarityRandom based on MountsRarity data

sgade opened the pull request:
Implements rarity-based randomization of the `MountList` as described in https://github.com/xod-wow/LiteMount/issues/60.

This requires the [MountsRarity](https://github.com/sgade/MountsRarity) addon which provides actual rarity data of mounts.

sgade commented on MountList.lua:
```diff
@@ -191,6 +191,33 @@ function LM.MountList:PriorityRandom(r)
+    local totalWeight = 0
+
+    for _, m in ipairs(self) do
+        local rarity = MountsRarityAddon.MountsRarity[tostring(m.mountID)]
```
I thought about moving this into a `LM.Mount:GetRarity()` function but then we'd have to reference to `MountsRarityAddon.MountsRarity` in two places (because we check for the early return above).

mbattersby replied:
I'm not worried about that, I'd prefer to eliminate the memory usage of the mountsRarity table as per the longer comment.

mbattersby replied:
If the LibStub used a funtion you could put the or 50 into that.

Alernatively your addon could metatable the data table to return 50 for not-found keys.

sgade replied:
I changed this to `local rarity = MountsRarity:GetRarityByID(m.mountID) or 50` in the next update.

sgade commented on MountList.lua:
```diff
@@ -191,6 +191,37 @@ function LM.MountList:PriorityRandom(r)
+        end
+        local invertedRarity = ( 1 / rarity )
+
+        mountsRarity[m.mountID] = invertedRarity
```
Thinking about it more, this is probably the place where we should also check for priority `0` (aka "disabled").
Makes even more sense to unify the two implementations.

mbattersby replied:
Yeah you can just set invertedRarity = 0 if m:GetPriority() == 0 I think.

sgade replied:
When the priority is set to `0`, it will not be in the list. So this is actually not needed.

mbattersby commented:
I thought about this a little more .

I wonder whether you would be happy to make your data addon into a LibStub library?

That way the packager can included a bundled version into LiteMount and people can separately install it if they want the super updated data. And LiteMount will get an updated version each time I do a release.

I think the change for your addon would be

```lua
local MAJOR, MINOR = "MountRarities-1.0", 1
local lib = LibStub:NewLibrary(MAJOR, MINOR)

lib.MountsRarity = {
   ...
}
```

and then it can be added into the pkgmeta.yaml and embeds.xml and MountList.lua can have

``
local LMR = LibStub("LibMountRarities-1.0")
```

You could implement GetRarityByID or even GetRarityWeightByID functions if you prefered rather than LiteMount reading the data directly, but not required.

Regarding the commonality between the rarity functions, I'm not worried about that. Please continue to keep RarityRandom separate even if you have to cut-and-paste all of PriorityRandom. It's easy to merge later when it's working but much harder to sepatate.

```lua
        local rarity = MountsRarityAddon.MountsRarity[tostring(m.mountID)]
```

This is a bit ugly, is it possible for you to index your data by the number rather than the string? I'm also a bit nervous about relying on tostring(nil) => nil and table[nil] => nil (since not all mounts have a .mountID) but it's probably ok.

I'm not sure the debug is warranted given that it will always print for non-journal mounts. I'd be happy with

``
local rarity = MountsRarityAddon.MountsRarity[m.mountID] or 50
```

You also don't need to create a mountsRarity table like you are, which is needed in PriorityRandom because of its bucketing system. It's just wasting memory. You can leave out `mountsRarity[m.mountID] = invertedRarity` and do `searchWeight = searchWeight + 1/MountsRarityAddon.MountsRarity[id]`.

Generally speaking I'm happy where this is going.

mbattersby commented:
I should also comment that before I merge this I'd probably want
1. you to have played around with the weights so they aren't so lopsided as previously discussed
2. some agreement from dataforazeroth that they are happy with what you are doing, since they are the data owners and I would hate to do all this work and then have them pull the plug on you

mbattersby commented on LiteMount.toc:
```diff
@@ -4,7 +4,7 @@
 ## Email: mib@post.com
 ## IconTexture: 132247
-## OptionalDependencies: LibDebug
+## OptionalDependencies: LibDebug, MountsRarity
```
We won't need this if we go down the LibStub route.

mbattersby commented on MountList.lua:
```diff
@@ -191,6 +191,33 @@ function LM.MountList:PriorityRandom(r)
 
 function LM.MountList:RarityRandom(r)
+    if #self == 0 then return end
+    if not MountsRarityAddon or not MountsRarityAddon.MountsRarity then
+        LM.Debug(" - RarityRandom: MountsRarity not loaded. Defaulting to SimpleRandom.")
+        return self:SimpleRandom(r)
+    end
```
Again if we go down the LibStub library route we won't need to worry about this as LiteMount will always have a copy.

sgade commented:
Thanks for your review. Will have time to implement your suggested changes in the coming days.

sgade commented:
I got the "ok" for the data use from the site's author. Via discord, so I'm not sure how to document that here. I will add notification so that if the API changes, I can adjust the generator with fixes. So that should be sorted out.

In other news, I'm still rebuilding the addon as a LibStub library. I thought it would be pretty straight forward but for some reason I can't get it to load the table from another addon initially. When defined like the following, it always returns an empty table at the start:
```lua
local MountsRarity = LibStub:NewLibrary("MountsRarity-2.0", 1) -- TODO: automatically upgrade the minor version on changes
if not MountsRarity then return end -- already loaded and no upgrade necessary

MountsRarity.data = {
  [6] = 64.33968131844426,
  [9] = 67.49980226212134,
  ...
}
```

Trying:
```
/script print(table.getn(LibStub("MountsRarity-2.0").data))
0
```

It works when called later from a slash command. Is there a way around that?

mbattersby commented:
> I got the "ok" for the data use from the site's author. Via discord, so I'm not sure how to document that here.

That's fine. I don't need any proof, I trust you. I just wanted to make sure you'd asked.

> I thought it would be pretty straight forward but for some reason I can't get it to load the table from another addon initially.

I think he problem here is that the addon isn't loaded before LiteMount is, which wouldn't happen with an embedded copy, so it's plain not there at that point. You may as well go ahead and put it back in the OptionalDependencies in the .toc, hopefully that will fix it up. I've never done this before either!

sgade commented:
I think I got it working almost the way you wished. There are two questions left that I cannot answer myself:

1. A UI element is needed so that the user can change the selected random function.
2. The weights are still lopsided. I tested that with my mounts and it is actually what I would expect. If I really want to sort my mounts by rarity, I want to see my super-rare mounts very often (because I like it, because I want to impress other players, ...). So the psychology works for me. Maybe you can try this out and see if it is the same for you?

mbattersby commented:
Hi, great stuff. I've merged this into a branch temporarily to have a play around with it. I hope to have some time to think about the weights and the UI on Sunday.

mbattersby commented:
Hi again, sorry to be a pain but I'll need you to specify a license for MountsRarity-2.0 so that I'm legally allowed to embed it.

Most embeddable libraries use a BSD-style license, e.g. https://github.com/SafeteeWoW/LibDeflate/blob/main/LICENSE.txt

But anything that is compatible with GPL and allows me to redistribute it is probably fine.

mbattersby commented:
Also I assume at some point you'll remove this
> print("no")
from the loader :)

mbattersby commented:
I unexpectedly had some time and have integrated your work into the feature_mountsrarity branch, including a UI option for enabling the rarity and some display of the rarity information. There's still some fine tuning needed and the texts are just placeholders but you should be able to test it out.

Totally agree 1/r weights is fine, no need to do anything.

sgade commented:
Thank you! I have added a license and removed the `print("no")` from the data addon. You probably want to update the committed files for the addon in LiteMount, now that it includes a license.

mbattersby commented:
I removed your committed Lib from my branch, the bigwigs packager will pull in the latest version from your repo (via the pkgmeta) every time I do a release.

mbattersby commented:
Please let me know how you get on with testing it out and if you have any suggestions. I expect I will probably run it a couple of weeks just to catch any bugs before I do a release. I'll also have to update the documentation and try to get some new strings translated.

sgade commented:
> I removed your committed Lib from my branch, the bigwigs packager will pull in the latest version from your repo (via the pkgmeta) every time I do a release.

Sounds good to me. Development and testing will require the addon be installed separately then, otherwise loading will fail.

> Please let me know how you get on with testing it out and if you have any suggestions.

I am cloning the branch to my main WoW install right now and will get back to you to discuss anything that comes up for me.

> try to get some new strings translated.

I can help with `deDE.`

mbattersby commented:
> Sounds good to me. Development and testing will require the addon be installed separately then, otherwise loading will fail.

I'm not sure how most addon developers do this to be honest. There is a shell script in Tools/fetch-externals.sh that I use to pull down the libs the same as the packager would. I assume there is something similar for other development environments (I use cygwin on windows because I'm an old school linux person). If you find anything out about how normal people do this please let me know. It must be a problem everyone has.

> I can help with `deDE`.

I haven't added any new strings for the rarity yet, but the deDE translations are here on curseforge:

https://legacy.curseforge.com/wow/addons/litemount/localization/languages/94/phrases#434541

sgade commented:
Now that LM does not ship a default copy of the rarity data, it should probably reference `MountsRarity` as an optional dependency. I have both installed but the order seems off now:
```lua
 Interface/AddOns/LiteMount/Mount.lua:17: Cannot find a library instance of "MountsRarity-2.0".
```

mbattersby commented:
Nah it needs the embedded version but (for release) it's pulled in by the bigwigs packager. For a development copy you'll need to pull them in yourself into Libs as per the "externals" in the pkgmeta. See my note above about the shell scripts I use.

mbattersby commented:
https://github.com/BigWigsMods/packager

This is what my CI uses to do the packaging. It reads the pkgmeta.yaml for control. It pulls in the externals as described in the pkgmeta before zipping up the release.

sgade commented:
I saw that and I think in that regard you shell script might just do duplicate work. You might want to think about doing the same thing for all dependencies, including Ace.
Pulling in all dependencies by hand is a bit tedious but the script might be helpful for that. But I think keeping a default version with the makes it much easier to get a development version working.

mbattersby commented:
I disagree, but no matter it's not changing. This is how everyone does it.

mbattersby commented:
Did you mean the script pull in all dependencies like Ace with the packager? It does. There are _no_ copied libs in LiteMount code any more.

sgade commented:
I saw BigWigs itself also does a mix so that might be considered a best-practise.

sgade commented:
I thought about using the script only after cloning the repo, so that new clones get the dependencies.

If you want to remove something from the repo but not delete it for you locally, you should be able to add it to `.gitignore` and to `git rm --cached <files>`.
